src/advanced/merkelized-map.md

```diff
@@ -0,0 +1,233 @@
+
+### Map With More Than One Element
+
+As defined above, in a map with 2 or more elements the MPT root always
+has two children. The root hash is defined as
+
+```text
+root_hash = hash(
+    HashTag::MapBranchNode ||
+    left_path || right_path
+    left_hash || right_hash
+).
+```
+
+Here
+
+- `left_path` and `right_path` are serialized paths to the child nodes;
+  their serialization format is described below.
+- `left_hash` and `right_hash` are 32-byte hashes associated with the
+  child nodes.
+
+Paths are serialized in this case as
+
+```text
+LEB128(bit_length) || bytes,
+```
+
+where
+
+- [LEB128] is a compact serialization format for unsigned integers
+- `bit_length` is the number of bits in the path
+- `bytes` is the path serialized as the minimum necessary number of bytes,
+  with zero padding if necessary.
+
+Hashes associated with child nodes are computed recursively:
+
+- If a node has 2 children, its hash is computed in the same way
+  as for the root
+- If a node is a leaf, its hash is `hash(HashTag::Blob || value)`,
+  where `value` is binary serialization of the value associated with the node.
+
+## MPT Proofs
+
+**Proofs** allow to efficiently and compactly verify that one or more
+values are associated with specific keys in a Merkelized map,
+or that a certain key has no associated value.
+
+One could think of proofs as of MPT pruning. That is, a proof is
+produced by “collapsing” some intermediate nodes in the MPT.
+Another point of view – from the light client perspective – is that a proof
+is essentially a limited view of a map, for which the MPT is
+constructed. This view allows to calculate the map hash and
+contains some of its elements.
+
+### Proof Example
+
+![Example of a Proof](../images/map-proof.png)
+
+This figure illustrates a map proof for the 4th element in a map
+with 7 elements. Intermediate MPT nodes included in the proof are highlighted.
+Pruned nodes are faded and have dashed contours.
+
+### Format
+
+The proofs returned by the Exonum storage engine are non-recursive,
+thus minimizing overhead (i.e., heap allocations) and allowing to use
+them in environments not allowing recursive data types. A proof
+consists of three principal parts:
+
+- Entries proven to exist in the map (i.g., values together with respective keys)
+- Keys proven to not exist in the map
+- Intermediate MPT nodes that together with entries allow to restore map hash
+
+??? note "Protobuf spec"
+    ```protobuf
+    message MapProof {
+      repeated OptionalEntry entries = 1;
+      repeated MapProofEntry proof = 2;
+    }
+
+    // Key with corresponding value and empty value if key is missing.
+    message OptionalEntry {
+      bytes key = 1;
+      oneof maybe_value {
+        bytes value = 2;
+        google.protobuf.Empty no_value = 3;
+      }
+    }
+
+    // Proof path and corresponding hash value.
+    message MapProofEntry {
+      // `proof_path` 34-byte serialization; see MerkleDB docs
```

Comment:
Don't quite like this either.

Reply:
If we will always use a compressed form for hashing, I think it would worth storing entries in compressed form as well.